reading_comprehension/qanet.py

```diff
@@ -0,0 +1,256 @@
+        The number of highway layers to use in between embedding the input and passing it through
+        the phrase layer.
+    phrase_layer : ``Seq2SeqEncoder``
+        The encoder (with its own internal stacking) that we will use in between embedding tokens
+        and doing the passage-question attention.
+    matrix_attention_layer : ``MatrixAttention``
+        The matrix attention function that we will use when comparing encoded passage and question
+        representations.
+    modeling_layer : ``Seq2SeqEncoder``
+        The encoder (with its own internal stacking) that we will use in between the bidirectional
+        attention and predicting span start and end.
+    dropout_prob : ``float``, optional (default=0.1)
+        If greater than 0, we will apply dropout with this probability between layers.
+    initializer : ``InitializerApplicator``, optional (default=``InitializerApplicator()``)
+        Used to initialize the model parameters.
+    regularizer : ``RegularizerApplicator``, optional (default=``None``)
+        If provided, will be used to calculate the regularization penalty during training.
+    """
+
+    def __init__(self, vocab: Vocabulary,
+                 text_field_embedder: TextFieldEmbedder,
+                 num_highway_layers: int,
+                 phrase_layer: Seq2SeqEncoder,
+                 matrix_attention_layer: MatrixAttention,
+                 modeling_layer: Seq2SeqEncoder,
+                 dropout_prob: float = 0.1,
+                 initializer: InitializerApplicator = InitializerApplicator(),
+                 regularizer: Optional[RegularizerApplicator] = None) -> None:
+        super(QANet, self).__init__(vocab, regularizer)
+
+        text_embed_dim = text_field_embedder.get_output_dim()
+        encoding_in_dim = phrase_layer.get_input_dim()
+        encoding_out_dim = phrase_layer.get_output_dim()
+        modeling_in_dim = modeling_layer.get_input_dim()
+        modeling_out_dim = modeling_layer.get_output_dim()
+
+        self._text_field_embedder = text_field_embedder
+
+        self._embedding_proj_layer = torch.nn.Linear(text_embed_dim, encoding_in_dim)
+        self._highway_layer = Highway(encoding_in_dim, num_highway_layers)
+
+        self._encoding_proj_layer = torch.nn.Linear(encoding_in_dim, encoding_in_dim)
+        self._phrase_layer = phrase_layer
+
+        self._matrix_attention = matrix_attention_layer
+
+        self._modeling_proj_layer = torch.nn.Linear(encoding_out_dim * 4, modeling_in_dim)
+        self._modeling_layer = modeling_layer
+
+        self._span_start_predictor = torch.nn.Linear(modeling_out_dim * 2, 1)
+        self._span_end_predictor = torch.nn.Linear(modeling_out_dim * 2, 1)
+
+        self._span_start_accuracy = CategoricalAccuracy()
+        self._span_end_accuracy = CategoricalAccuracy()
+        self._span_accuracy = BooleanAccuracy()
+        self._squad_metrics = SquadEmAndF1()
+        self._dropout = torch.nn.Dropout(p=dropout_prob)
+
```

Comment:
Is it worth having any consistency checks on the dimensions of your inputs?  I'm thinking of something like this: https://github.com/allenai/allennlp/blob/aa1b774ed8de31ec04bebf9f054200bc2507e0c5/allennlp/models/reading_comprehension/bidaf.py#L96-L103